elfo-network/src/socket/capabilities/compression.rs

```diff
@@ -0,0 +1,157 @@
+// Layouts are specified from highest to lowest bits.
+
+use std::fmt;
+
+use crate::config::Preference;
+
+bitflags::bitflags! {
+    /// Set of algorithms. 24 bits.
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    pub(crate) struct Algorithms: u32 {
+        const LZ4 = 1;
+        // NB: Shift by 2: `const ZSTD = 1 << 2;`.
+    }
+}
+
+/// Compression capabilities.
+///
+/// Layout:
+/// ```text
+///       22        2
+/// ┌────────────┬─────┐
+/// │  Reserved  │ Lz4 │
+/// └────────────┴─────┘
+/// ```
+///
+/// Each mentioned algorithm here occupies two bits for a reason, here's the
+/// layout of those bits:
+/// ```text
+///       1           1
+/// ┌───────────┬───────────┐
+/// │ Preferred │ Supported │
+/// └───────────┴───────────┘
+/// ```
+///
+/// 1. Preferred - the compression algorithm is preferred, implies `Supported`.
+/// 2. Supported - the compression algorithm is supported.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Compression(u32);
+
+impl Compression {
+    pub(crate) const fn empty() -> Self {
+        Self::new(Algorithms::empty(), Algorithms::empty())
+    }
+
+    pub(super) const fn from_bits_unchecked(v: u32) -> Self {
+        Self(v)
+    }
+
+    pub(crate) const fn from_bits_truncate(v: u32) -> Self {
+        let supported = Algorithms::from_bits_truncate(v);
+        let preferred = Algorithms::from_bits_truncate(v >> 1);
+
+        Self::new(supported, preferred)
+    }
+
+    pub(crate) const fn new(supported: Algorithms, preferred: Algorithms) -> Self {
+        let preferred = preferred.bits();
+        // Preferred implies supported.
+        let supported = supported.bits() | preferred;
+
+        // 0 1 0 1 | Supported
+        // 1 0 1 0 | Preferred
+        // -------
+        // 1 1 1 1
+        let joined = supported | (preferred << 1);
+
+        Self(joined)
+    }
+
+    pub(crate) fn toggle(&mut self, algos: Algorithms, pref: Preference) {
```

Comment:
Actually, It's not a "toggle", because this code does not work as expected for `toggle(Preferred); toggle(Disabled)` sequence.